lib/kafka-producer.js

```diff
@@ -70,41 +70,40 @@ Producer.prototype.getProducer = Promise.method(function (opts, topicOptions) {
  * @private
  */
 Producer.prototype._serializeType = function (topicName, isKey, data) {
-
   let schema = null;
   let subject = null;
 
-  if (isKey === false) {
-    subject = this.sr.valueSubjectStrategy.prepareSubjectName(topicName, data);
-    schema = this.sr.valueSchemas[subject];
-  } else {
+  if (isKey === true) {
     subject = this.sr.keySubjectStrategy.prepareSubjectName(topicName, data);
-    schema = this.sr.keySchemas[subject];
-  }
+    schema = this.sr.schemas[subject];
+  } else {
+    subject = this.sr.valueSubjectStrategy.prepareSubjectName(topicName, data);
+    schema = this.sr.schemas[subject];
 
-  const schemaType = isKey === false
-    ? 'value'
-    : 'key';
+  }
 
   if (!schema) {
     log.warn('_produceWrapper() :: Warning, did not find topic on SR for subject schema:',
       {
         topic: topicName,
-        subject: subject,
-        schemaType,
+        subject: subject
       });
 
     if (this.__shouldFailWhenSchemaIsMissing) {
       throw new Error(
-        'Unable to serialize message for topic ' + topicName
-        + ' and type ' + schemaType + ': schema not found');
+        `Unable to serialize message for topic ${topicName} : schema not found`);
     }
 
-    return (typeof data === 'object')
-      ? new Buffer(JSON.stringify(data))
-      : new Buffer(data);
+    if (typeof data === 'object') {
+      // Making sure __schemaName and __id is not serialized in case avro schema is missing and json is used as a fallback.
+      delete data.__schemaName;
+      delete data.__id;
+      return new Buffer(JSON.stringify(data));
+    } else {
+      return new Buffer(data);
+    }
   } else {
-    const schemaId = this.sr.schemaIds[subject + '-' + schemaType];
+    const schemaId = data.__id || this.sr.schemaIds[subject];
```

Comment:
if __id is present on the object to be serialized use that, otherwise fallback to this.sr.schemaIds[subject]